crates/typst/src/eval/call.rs

```diff
@@ -129,23 +128,20 @@ impl Eval for ast::FuncCall<'_> {
             (callee.eval(vm)?, args.eval(vm)?.spanned(span))
         };
 
-        // Handle math special cases for non-functions:
-        // Combining accent symbols apply themselves while everything else
-        // simply displays the arguments verbatim.
-        if in_math && !matches!(callee, Value::Func(_)) {
-            if let Value::Symbol(sym) = &callee {
-                let c = sym.get();
-                if let Some(accent) = Symbol::combining_accent(c) {
-                    let base = args.expect("base")?;
-                    let size = args.named("size")?;
-                    args.finish()?;
-                    let mut accent = AccentElem::new(base, Accent::new(accent));
-                    if let Some(size) = size {
-                        accent = accent.with_size(size);
-                    }
-                    return Ok(Value::Content(accent.pack()));
+        // If the callee is a callable symbol, call its handler.
+        let callee = match callee {
+            Value::Symbol(sym) => {
+                if let Some(handler) = sym.func() {
+                    Value::Func(handler)
+                } else {
+                    Value::Symbol(sym)
                 }
             }
+            callee => callee,
+        };
+
+        // For non-functions in math, we wrap the arguments in parentheses.
+        if in_math && !matches!(callee, Value::Func(_)) {
```

Comment:
We could try doing the `let callee = callee.cast::<Func>().at(callee_span);` above here, but not yet early return if it's `Err(_)`, instead first checking the non-function in math case.

Then, we could add `Symbol -> Func` to the cast implementation of `Func` in `foundations/value.rs` (like `Type` has). and get rid of the symbol special case above.

Not 100% whether it'll work, but worth a try.

Reply:
I tried it just now. If we simply do something like:

```rs
primitive! {
    // ...
    // NOTE: `try_func()` wraps `func()` to give `StrResult`.
    // The error message is "symbol {} is not callable"
    Symbol(symbol) => symbol.try_func()?,
}
```

It cannot pass the tests. It seems because the compiler tries to cast some symbols into function, even if the symbol is not called. For example, when the following typst file gets compiled, the compiler emits an error: "symbol ∫ is not callable".

```typ
$ ∫_a^b iota_a^b $
```

To implement the cast without affecting current behavior of the compiler, I tried to add a `if` guard to the `primitive` macro, so that `Symbol -> Func` is not always castable:

```rs
primitive! {
    // ...
    Symbol(symbol) if symbol.func().is_some() => symbol.func().unwrap(),
}
```

I will push the code soon, and you can see my implementation.

Reply:
Sorry for my mistake. The full case is:

```typ
#let eq = $ ∫_a^b iota_a^b $
#eq
#show "∫": math.limits
#show math.iota: math.limits.with(inline: false)
#eq
$iota_a^b$
```

It seems the show rule results in the cast

Reply:
Here is the minimal case:

```typ
#show math.iota: math.limits.with(inline: false)
$ iota_a^b $
```

Reply:
The stack backtrace indicates that `Selector` tries to cast anything to `Func`, so this is the reason why `Symbol -> Func` is applied in this case.

```txt
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_display
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:197:5
   3: core::panicking::panic_str
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:172:5
   4: core::option::expect_failed
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/option.rs:1995:5
   5: core::option::Option<T>::expect
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/option.rs:896:21
   6: typst::foundations::value::<impl typst::foundations::cast::FromValue for typst::foundations::func::Func>::from_value
             at ./crates/typst/src/foundations/value.rs:665:23
   7: <typst::foundations::selector::Selector as typst::foundations::cast::FromValue>::from_value
             at ./crates/typst/src/foundations/selector.rs:278:1
   8: <typst::foundations::selector::ShowableSelector as typst::foundations::cast::FromValue>::from_value
             at ./crates/typst/src/foundations/selector.rs:450:24
   9: typst::foundations::value::Value::cast
             at ./crates/typst/src/foundations/value.rs:155:9
  10: typst::eval::rules::<impl typst::eval::Eval for typst_syntax::ast::ShowRule>::eval::{{closure}}
             at ./crates/typst/src/eval/rules.rs:37:24
  11: core::option::Option<T>::map
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/option.rs:1073:29
  12: typst::eval::rules::<impl typst::eval::Eval for typst_syntax::ast::ShowRule>::eval
             at ./crates/typst/src/eval/rules.rs:35:24
  13: typst::eval::markup::eval_markup
             at ./crates/typst/src/eval/markup.rs:42:30
  14: typst::eval::markup::<impl typst::eval::Eval for typst_syntax::ast::Markup>::eval
             at ./crates/typst/src/eval/markup.rs:19:9
  15: typst::eval::eval::{{closure}}
             at ./crates/typst/src/eval/mod.rs:74:18
```

Reply:
I see. In that case, you can move the `text: EcoString` arm upwards to prefer casting a symbol to a text selector over a function selector (in particular since symbols won't generate element functions anyway).

Reply:
Thanks! Now it passes all the tests :heart: